### PR TITLE
Implement Browser-Restricted Login Codes using `has_secure_token`

### PR DESCRIPTION
**Pull request by sampoder**
When using an email login code from another browser, you will now get an error
image

---

This only works for Email login codes, and not SMS since we do not have a model for SMS to store a browser token.

### CODE REVIEW

**Comment by sampoder** [comment]
This supersedes #4220

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -72,6 +72,9 @@ def login_code
       flash[:error] = resp[:error]
       return redirect_to auth_users_path
     end
+
+    cookies.signed[:browser_token] = { value: resp[:browser_token], expires: Time.current + 15.minutes }
```
Comment (garyhtou) [inline app/controllers/users_controller.rb]:
```suggestion
    cookies.signed[:browser_token] = { value: resp[:browser_token], expires: 15.minutes.from_now }
```

File: app/services/user_service/exchange_login_code_for_user.rb [hunk 2]
```diff
@@ -36,6 +37,7 @@ def exchange_login_code_by_email
 
       raise ::Errors::InvalidLoginCode if login_code.nil?
       raise ::Errors::InvalidLoginCode if login_code.created_at < (Time.current - 15.minutes)
+      raise ::Errors::BrowserMismatch if login_code.browser_token.present? && (login_code.browser_token != @browser_token || !@browser_token)
```
Comment (garyhtou) [inline app/services/user_service/exchange_login_code_for_user.rb]:
```suggestion
      raise ::Errors::BrowserMismatch if login_code.browser_token.present? && (login_code.browser_token != @browser_token)
```

File: app/services/user_service/exchange_login_code_for_user.rb [hunk 3]
```diff
@@ -2,10 +2,11 @@
 
 module UserService
   class ExchangeLoginCodeForUser
-    def initialize(user_id:, login_code:, sms: false)
+    def initialize(user_id:, login_code:, sms: false, browser_token: "")
```
Comment (garyhtou) [inline app/services/user_service/exchange_login_code_for_user.rb]:
```suggestion
    def initialize(user_id:, login_code:, sms: false, browser_token:)
```
let's make this required. you might need to update one of the specs

File: app/services/user_service/exchange_login_code_for_user.rb (hunk 2 above)
Comment (garyhtou) [reply app/services/user_service/exchange_login_code_for_user.rb]: I think that check might not be necessary

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -73,6 +73,9 @@ def login_code
       flash[:error] = resp[:error]
       return redirect_to auth_users_path
     end
+
+    cookies.signed[:browser_token] = { value: resp[:browser_token], expires: 15.minutes.from_now }
```
Comment (garyhtou) [inline app/controllers/users_controller.rb]:
```suggestion
    cookies.signed[:browser_token] = { value: resp[:browser_token], expires: LoginCode::EXPIRATION.from_now }
```

File: app/services/user_service/exchange_login_code_for_user.rb [hunk 5]
```diff
@@ -36,6 +37,7 @@ def exchange_login_code_by_email
 
       raise ::Errors::InvalidLoginCode if login_code.nil?
       raise ::Errors::InvalidLoginCode if login_code.created_at < (Time.current - 15.minutes)
+      raise ::Errors::BrowserMismatch if login_code.browser_token.present? && (login_code.browser_token != @browser_token)
```
Comment (garyhtou) [inline app/services/user_service/exchange_login_code_for_user.rb]: Let's use https://api.rubyonrails.org/classes/ActiveSupport/SecurityUtils.html#method-c-secure_compare

File: app/services/login_code_service/request.rb [hunk 6]
```diff
@@ -47,7 +47,8 @@ def send_login_code_by_email(user)
       {
         id: user.id,
         email: user.email,
-        status: "login code sent"
+        status: "login code sent",
+        browser_token: login_code.browser_token
```
Comment (garyhtou) [inline app/services/login_code_service/request.rb]:
```suggestion
        browser_token: login_code.browser_token,
        login_code:
```

File: app/controllers/users_controller.rb (hunk 4 above)
Comment (garyhtou) [inline app/controllers/users_controller.rb]:
```suggestion
    cookies.signed[:"browser_token_#{resp[:login_code].id}"] = { value: resp[:browser_token], expires: 15.minutes.from_now }
```

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -178,6 +182,14 @@ def exchange_login_code
     @force_use_email = params[:force_use_email]
     initialize_sms_params
     return render :login_code, status: :unprocessable_entity
+  rescue Errors::BrowserMismatch => e
```
Comment (garyhtou) [inline app/controllers/users_controller.rb]: Combine this with the rescue above:

```ruby
rescue Errors::InvalidLoginCode, Errors::BrowserMismatch => e
  message = case e
                      when Errors::InvalidLoginCode
											  "Invalid login code!"
                      when Errors::BrowserMismatch
											  "Looks like this isn't the browser that requested that code!"
                      end

    flash.now[:error] = message
    # Propagate the to the login_code page on invalid code
    @user_id = params[:user_id]
    @email = params[:email]
    @force_use_email = params[:force_use_email]
    initialize_sms_params
    return render :login_code, status: :unprocessable_entity
end  
```

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -157,7 +160,8 @@ def exchange_login_code
     user = UserService::ExchangeLoginCodeForUser.new(
       user_id: params[:user_id],
       login_code: params[:login_code],
-      sms: params[:sms]
+      sms: params[:sms],
+      browser_token: cookies.signed[:browser_token]
```
Comment (sampoder) [inline app/controllers/users_controller.rb]:
```suggestion
      browser_token: cookies.signed[:"browser_token_#{resp[:login_code].id}"]
```

File: app/services/user_service/exchange_login_code_for_user.rb (hunk 5 above)
Comment (sampoder) [reply app/services/user_service/exchange_login_code_for_user.rb]:
```suggestion
      raise ::Errors::BrowserMismatch if login_code.browser_token.present? && !ActiveSupport::SecurityUtils.secure_compare(login_code.browser_token, @browser_token)
```

**Review by garyhtou — APPROVED** [review]
Good to go, just fix the rspec failure before merging

File: spec/models/login_code_spec.rb [hunk 9]
```diff
@@ -53,6 +53,7 @@
     it "retrys when colliding with an existing login code" do
       create(:login_code, code: "123456")
 
+      allow(SecureRandom).to receive(:random_number).with(58).and_call_original
```
Comment (cjdenio) [inline spec/models/login_code_spec.rb]: What's this for?

Comment (cjdenio) [reply spec/models/login_code_spec.rb]: ah, fascinating
https://github.com/rails/rails/blame/3064d4f53a48beb107a549a2249820350e698050/activesupport/lib/active_support/core_ext/securerandom.rb#L27

Comment (sampoder) [reply spec/models/login_code_spec.rb]: yep!

**Pull request merged**